Sources/NIOHTTP1/HTTPTypes.swift

```diff
@@ -174,17 +174,323 @@ struct HTTPHeader {
     let value: HTTPHeaderIndex
 }
 
+
+private extension UInt8 {
+    var isASCII: Bool {
+        return self <= 127
+    }
+
+    var maskedLowercase: UInt8 {
+        return self & 0xdf
+    }
+}
+
+
+/// Known HTTP header name. These should be created once and re-used as instances are memory intensive and "expensive" to create.
+public struct HTTPHeaderName {
+
+    /// Box for different fields to reduce reference count overhead.
+    private final class _Storage {
+        /// Holds the bytes of the name + ': '
+        let bytesWithSeparator: ContiguousArray<UInt8>
+        /// Holds the lowercased bases (masked) of the name.
+        let lowerCaseBytes: ContiguousArray<UInt8>
+
+        let lowerCaseName: String
+
+        init(_ value: String) {
+            precondition(!value.utf8.contains(where: { !$0.isASCII }), "name must be ASCII")
+            var array = ContiguousArray(value.utf8)
+            array.append(contentsOf: ": ".utf8)
+            self.bytesWithSeparator = array
+            self.lowerCaseBytes = ContiguousArray(value.utf8.map { $0.maskedLowercase })
```

Comment:
@gottesmm and this I would've hoped we can hold as a stored property that returns `string.utf8.lazy.map { $0.maskedLowercase }` so no extra storage needed.

Reply:
actually, if we do store lower case, we can just use `memcmp`

Reply:
@normanmaurer I know it's only invoked very few times but you're missing a `.lazy` here. Also this would be quicker:

    self.lowerCaseBytes = ContiguousArray(self.bytesWithSeparator.lazy.dropLast(2).map { $0.maskedLowercase })

Reply:
this is also for @milseman